Separate heartbeat failure from cleanup in TxnCoordSender.

bdarnell opened the pull request:
If a heartbeat fails, stop heartbeating but do not perform
other cleanup. The client doesn't know the heartbeat
has failed and so will still try to complete the transaction,
and it needs to get a retryable error instead of the util.Errorf
that would result if the heartbeat removed the transaction.

Closes #2028.

bdarnell committed:
Separate heartbeat failure from cleanup in TxnCoordSender.

If a heartbeat fails, stop heartbeating but do not perform
other cleanup. The client doesn't know the heartbeat
has failed and so will still try to complete the transaction,
and it needs to get a retryable error instead of the util.Errorf
that would result if the heartbeat removed the transaction.

Closes #2028.

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -802,3 +783,61 @@ func (tc *TxnCoordSender) heartbeat(id string) {
+	txn := txnMeta.txn
+	tc.Unlock()
+	if !proceed {
+		tc.unregisterTxn(id)
```
This method is mostly just code movement, but there are two changes. This line was added...

bdarnell commented on kv/txn_coord_sender.go:
```diff
@@ -802,3 +783,61 @@ func (tc *TxnCoordSender) heartbeat(id string) {
+	if reply.GoError() != nil {
+		log.Warningf("heartbeat to %s failed: %s", txn, reply.GoError())
+	} else if reply.Txn != nil && reply.Txn.Status != proto.PENDING {
+		// The transaction has been committed or aborted, so no use
```
And the unregisterTxn that used to happen here (via a defer) is removed.

tschottdorf replied:
you're going to leak entries in the transaction map with this, I think. The map entry and the heartbeat goroutine should only exist in parallel (hence the `defer` in the previous code).
From our discussion in #2128, I would have expected that you simply keep running on heartbeat failures (if status not PENDING, you could stop sending actual heartbeats, of course). The goroutine would then shut down with an explicit commit by the client or by being abandoned,

bdarnell replied:
Right. I put the `unregisterTxn` call back where it was, and removed the new test I added (which was getting to be tricky to synchronize with the black-box heartbeat loop).

bdarnell committed:
Fix leak of txnMeta records.

Remove the new test, which relied on events that are difficult to
observe in this implementation.

tschottdorf commented on kv/txn_coord_sender.go:
```diff
@@ -396,14 +396,26 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
 			}
 			tc.Lock()
-			if txnMeta := tc.txns[id]; id != "" && txnMeta != nil {
+			txnMeta, metaOK := tc.txns[id]
```
slight preference for
```go
txnMeta := tc.txns[id]
metaOK := txnMeta != nil // don't even really need this
```

since the current version has a (theoretically) unhandled case in which there is a `nil` entry in the map.

bdarnell replied:
If we ever insert a nil pointer in the map we have a leak, since nothing will ever clean up that map entry. I prefer to use the `, ok` lookup to test for existence; if we ever get a nil pointer here the panic will alert us to the fact that something is not as we expect.

tschottdorf replied:
ok.

tschottdorf commented on kv/txn_coord_sender.go:
```diff
@@ -668,15 +680,16 @@ func (tc *TxnCoordSender) updateResponseTxn(argsHeader *proto.RequestHeader, rep
 	tc.Lock()
 	defer tc.Unlock()
-	txnMeta, ok := tc.txns[string(txn.ID)]
+	txnMeta, ok := tc.txns[id]
```
ditto (though I likely wrote that code)

tschottdorf commented:
LGTM, but I'm not sure why you removed `TestTxnCoordIdempotentCleanup`. Has anything changed there or is it an accidental deletion?

bdarnell commented:
I had removed `TestTxnCoordIdempotentCleanup` because in the first iteration of this PR, cleanup was no longer idempotent. But now it's idempotent again so I've restored the test.

bdarnell committed:
Restore TestTxnCoordIdempotentCleanup

tamird commented on kv/txn_coord_sender.go:
```diff
@@ -396,14 +396,26 @@ func (tc *TxnCoordSender) sendOne(ctx context.Context, call proto.Call) {
+				// TODO(bdarnell): if we had a GetTransactionStatus API then
+				// we could lookup the transaction and return either nil or
+				// TransactionAbortedError instead of this ambivalent error.
+				call.Reply.Header().SetGoError(util.Errorf(
```
this is the cause of the regression in https://github.com/cockroachdb/cockroach/issues/2145. I believe there's a missing `return` after this line.

tschottdorf replied:
good catch. Those have happened before.

tschottdorf replied:
It will be really good if we manage to move to `(Request, Response)` throughout.
